src/Microsoft.OData.Core/Json/ODataJsonWriterUtilsAsync.cs

```diff
@@ -0,0 +1,274 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ODataJsonWriterUtilsAsync.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Json
+{
+    #region Namespaces
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.OData.JsonLight;
+    #endregion Namespaces
+
+    /// <summary>
+    /// Helper methods used by the OData writer for the JSON format.
+    /// </summary>
+    internal static partial class ODataJsonWriterUtils
+    {
+        /// <summary>
+        /// Asynchronously writes an error message.
+        /// </summary>
+        /// <param name="jsonWriter">The JSON writer to write the error.</param>
+        /// <param name="writeInstanceAnnotationsDelegate">Delegate to write the instance annotations.</param>
+        /// <param name="error">The error instance to write.</param>
+        /// <param name="includeDebugInformation">A flag indicating whether error details should be written (in debug mode only) or not.</param>
+        /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <param name="writingJsonLight">true if we're writing JSON lite, false if we're writing verbose JSON.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal static async Task WriteErrorAsync(IJsonWriterAsync jsonWriter,
+            Func<IEnumerable<ODataInstanceAnnotation>, Task> writeInstanceAnnotationsDelegate, ODataError error,
+            bool includeDebugInformation, int maxInnerErrorDepth, bool writingJsonLight)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter != null");
+            Debug.Assert(error != null, "error != null");
+
+            string code, message;
+            ErrorUtils.GetErrorDetails(error, out code, out message);
+
+            ODataInnerError innerError = includeDebugInformation ? error.InnerError : null;
+
+            await WriteErrorAsync(

```

Comment:
>await [](start = 12, length = 5)

pass-through where we can probably elide the await *provided we are confident that error.GetInstanceAnnotations() will never throw an exception*.

Reply:
@habbes FYI... This is what we were debating about in my previous PR. We can duck generation of the `async` state machine in quite a few places where the method is a passthrough and relieve the pressure on the GC in particular since that's where the higher cost it. I'll consider a PR to address the scenarios we had identified previously.

Reply:
@gathogojr Yeah, I also think we should remove await from pass-through methods